commons/src/main/java/io/aiven/kafka/tieredstorage/commons/UniversalRemoteStorageManager.java

```diff
@@ -16,54 +16,263 @@
+            if (logSegmentData.transactionIndex().isPresent()) {
+                uploadIndexFile(remoteLogSegmentMetadata, Files.newInputStream(logSegmentData.transactionIndex().get()),
+                    TRANSACTION);
+            }
+            uploadIndexFile(remoteLogSegmentMetadata, new ByteBufferInputStream(logSegmentData.leaderEpochIndex()),
+                LEADER_EPOCH);
+        } catch (final IOException e) {
+            throw new RemoteStorageException(e);
+        }
+    }
+
+    private void uploadIndexFile(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                 final InputStream index,
+                                 final IndexType indexType) throws IOException {
+        objectStorageFactory.fileUploader().upload(index,
+            objectKey.key(remoteLogSegmentMetadata, ObjectKey.Suffix.fromIndexType(indexType)));
+    }
+
+    private void uploadManifest(final RemoteLogSegmentMetadata remoteLogSegmentMetadata,
+                                final SegmentManifestV1 segmentManifestV1)
+        throws IOException {
+        final String manifest = mapper.writeValueAsString(segmentManifestV1);
+        final String manifestFileKey =
+            objectKey.key(remoteLogSegmentMetadata, ObjectKey.Suffix.MANIFEST);
+
+        objectStorageFactory.fileUploader().upload(new ByteArrayInputStream(manifest.getBytes()), manifestFileKey);
+    }
+
+    protected static Cipher encryptionCipherSupplier(final EncryptionKeyAndAAD encryptionKeyAndAAD) {
+        try {
+            final Cipher encryptCipher = getCipher();
+            encryptCipher.init(Cipher.ENCRYPT_MODE, encryptionKeyAndAAD.key, SecureRandom.getInstanceStrong());
+            encryptCipher.updateAAD(encryptionKeyAndAAD.aad);
+            return encryptCipher;
+        } catch (final NoSuchAlgorithmException | InvalidKeyException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected static Cipher getCipher() {
+        try {
+            return Cipher.getInstance("AES/GCM/NoPadding", "BC");
+        } catch (final NoSuchAlgorithmException | NoSuchProviderException | NoSuchPaddingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    protected static Cipher decryptionCipherSupplier(final byte[] encryptedChunk,
+                                                     final SegmentEncryptionMetadata encryptionMetadata) {
+        try {
+            final Cipher encryptCipher = getCipher();
+            encryptCipher.init(Cipher.DECRYPT_MODE, encryptionMetadata.secretKey(),
+                new IvParameterSpec(encryptedChunk, 0, encryptionMetadata.ivSize()),
+                SecureRandom.getInstanceStrong());
+            encryptCipher.updateAAD(encryptionMetadata.aad());
+            return encryptCipher;
+        } catch (final NoSuchAlgorithmException | InvalidKeyException | InvalidAlgorithmParameterException e) {
+            throw new RuntimeException(e);
+        }
     }
```

Comment:
This doesn't belong here. Should we move it to `security.EncryptionKeyProvider`?

Reply:
Solved on #184